src/autopas/selectors/tuningStrategy/TuningStrategySuperClass.h

```diff
@@ -0,0 +1,130 @@
+ public:
+  /**
+   * Constructor for the TuningStrategySuperClass that generates the search space from the allowed options.
+   * @param allowedContainerOptions
+   * @param allowedTraversalOptions
+   * @param allowedDataLayoutOptions
+   * @param allowedNewton3Options
+   * @param allowedCellSizeFactors
+   */
+  TuningStrategySuperClass(const std::set<ContainerOption> &allowedContainerOptions,
+                           const std::set<double> &allowedCellSizeFactors,
+                           const std::set<TraversalOption> &allowedTraversalOptions,
+                           const std::set<DataLayoutOption> &allowedDataLayoutOptions,
+                           const std::set<Newton3Option> &allowedNewton3Options)
+      : _containerOptions(allowedContainerOptions) {
+    // sets search space and current config
+    populateSearchSpace(allowedContainerOptions, allowedCellSizeFactors, allowedTraversalOptions,
+                        allowedDataLayoutOptions, allowedNewton3Options);
+  }
+
+  /**
+   * Constructor for the TuningStrategySuperClass that only contains the given configurations.
+   * This constructor assumes only valid configurations are passed! Mainly for easier unit testing.
+   * @param allowedConfigurations Set of configurations AutoPas can choose from.
+   */
+  explicit TuningStrategySuperClass(std::set<Configuration> allowedConfigurations)
+      : _containerOptions{}, _searchSpace(std::move(allowedConfigurations)) {
+    for (const auto &config : _searchSpace) {
+      _containerOptions.insert(config.container);
+    }
+  }
+
+  inline std::set<ContainerOption> getAllowedContainerOptions() const override { return _containerOptions; }
+
+  inline bool searchSpaceIsTrivial() const override { return _searchSpace.size() == 1; }
+
+  inline bool searchSpaceIsEmpty() const override { return _searchSpace.empty(); }
+
+ protected:
+  /**
+   * Fills the search space with the cartesian product of the given options (minus invalid combinations).
+   * @param allowedContainerOptions
+   * @param allowedCellSizeFactors
+   * @param allowedTraversalOptions
+   * @param allowedDataLayoutOptions
+   * @param allowedNewton3Options
+   */
+  inline void populateSearchSpace(const std::set<ContainerOption> &allowedContainerOptions,
+                                  const std::set<double> &allowedCellSizeFactors,
+                                  const std::set<TraversalOption> &allowedTraversalOptions,
+                                  const std::set<DataLayoutOption> &allowedDataLayoutOptions,
+                                  const std::set<Newton3Option> &allowedNewton3Options);
+
+  /**
+   * Finds the optimum in an unordered map
+   * @param findOptimumSearchSpace
+   * @return optimum
+   */
+  static inline std::__detail::_Node_const_iterator<std::pair<const Configuration, unsigned long>, false, true>
+  getOptimum(const std::unordered_map<Configuration, size_t, ConfigHash> &findOptimumSearchSpace);
```

Comment:
this is a good candidate where the return type should just be `auto`

Reply:
If I use auto I get this error message in the building process and I have no clue why:
/usr/bin/ld: CMakeFiles/runTests.dir/tests/selectors/tuningStrategy/FullSearchTest.cpp.o: in function `autopas::TuningStrategySuperClass::getOptimum(std::unordered_map<autopas::Configuration, unsigned long, autopas::ConfigHash, std::equal_to<autopas::Configuration>, std::allocator<std::pair<autopas::Configuration const, unsigned long> > > const&)':
FullSearchTest.cpp:(.text+0x3a0): multiple definition of `autopas::TuningStrategySuperClass::getOptimum(std::unordered_map<autopas::Configuration, unsigned long, autopas::ConfigHash, std::equal_to<autopas::Configuration>, std::allocator<std::pair<autopas::Configuration const, unsigned long> > > const&)'; CMakeFiles/runTests.dir/tests/selectors/AutoTunerTest.cpp.o:AutoTunerTest.cpp:(.text+0x2240): first defined here
/usr/bin/ld: CMakeFiles/runTests.dir/tests/selectors/tuningStrategy/PredictiveTuningTest.cpp.o: in function `autopas::TuningStrategySuperClass::getOptimum(std::unordered_map<autopas::Configuration, unsigned long, autopas::ConfigHash, std::equal_to<autopas::Configuration>, std::allocator<std::pair<autopas::Configuration const, unsigned long> > > const&)':
PredictiveTuningTest.cpp:(.text+0x3a0): multiple definition of `autopas::TuningStrategySuperClass::getOptimum(std::unordered_map<autopas::Configuration, unsigned long, autopas::ConfigHash, std::equal_to<autopas::Configuration>, std::allocator<std::pair<autopas::Configuration const, unsigned long> > > const&)'; CMakeFiles/runTests.dir/tests/selectors/AutoTunerTest.cpp.o:AutoTunerTest.cpp:(.text+0x2240): first defined here
/usr/bin/ld: ../../src/autopas/libautopas.a(TuningStrategyFactory.cpp.o): in function `autopas::TuningStrategySuperClass::getOptimum(std::unordered_map<autopas::Configuration, unsigned long, autopas::ConfigHash, std::equal_to<autopas::Configuration>, std::allocator<std::pair<autopas::Configuration const, unsigned long> > > const&)':
TuningStrategyFactory.cpp:(.text+0x360): multiple definition of `autopas::TuningStrategySuperClass::getOptimum(std::unordered_map<autopas::Configuration, unsigned long, autopas::ConfigHash, std::equal_to<autopas::Configuration>, std::allocator<std::pair<autopas::Configuration const, unsigned long> > > const&)'; CMakeFiles/runTests.dir/tests/selectors/AutoTunerTest.cpp.o:AutoTunerTest.cpp:(.text+0x2240): first defined here
collect2: error: ld returned 1 exit status
make[2]: *** [tests/testAutopas/CMakeFiles/runTests.dir/build.make:1054: tests/testAutopas/runTests] Error 1
make[1]: *** [CMakeFiles/Makefile2:920: tests/testAutopas/CMakeFiles/runTests.dir/all] Error 2
make: *** [Makefile:112: all] Error 2

Reply:
Are you sure you changed the signature at bot the declaration and the implementation of the function?
If I change the declaration (around line 77) to:
```
static inline auto getOptimum( .... );
```
and the implementation (around line 122) to:
```
auto TuningStrategySuperClass::getOptimum( ...
```
it compiles on my machine.

Reply:
I forgot `inline`. Now it works.